api/include/opentelemetry/context/runtime_context.h

```diff
@@ -108,15 +146,46 @@ class RuntimeContext
     return temp_context.GetValue(key);
   }
 
-protected:
-  // Provides a token with the passed in context
-  Token CreateToken(Context context) noexcept { return Token(context); }
+  /**
+   * Provide a custom runtime context storage.
+   *
+   * This provides a possibility to override the default thread-local runtime
+   * context storage. This has to be set before any spans are created by the
+   * application, otherwise context state information will be discarded, which
+   * might lead to unexpected results.
+   *
+   * @param storage a custom runtime context storage
+   */
+  static void SetRuntimeContextStorage(nostd::shared_ptr<RuntimeContextStorage> storage) noexcept
+  {
+    while (GetLock().test_and_set(std::memory_order_acquire))
+      ;
+    GetStorage() = storage;
+    GetLock().clear(std::memory_order_release);
+  }
+
+private:
+  static nostd::shared_ptr<RuntimeContextStorage> GetRuntimeContextStorage() noexcept
+  {
+    while (GetLock().test_and_set(std::memory_order_acquire))
```

Comment:
I'm trying to understand the purpose of the lock here:

- line 178 - `&GetStorage()` is already atomic, since static at line 180 is `magic static` aka `function-local static`. Since C++11, the initialization of magic statics is guaranteed to be thread safe...

- whereas `storage` at line 172 is on stack...

Could you please clarify on what is the race we are trying to address? Is this a race between getting storage for the first time vs. `SetRuntimeContextStorage` potentially being called at the same time?

I was thinking about slightly less elegant but shorter solution.. It won't require a lock if we declare that the context storage customization MUST be done before any other telemetry API call:
https://github.com/maxgolov/opentelemetry-cpp/blob/4004c2d1ebc6bd9b1e55eaf9310726fc939f9de5/api/include/opentelemetry/context/runtime_context.h#L28

Reply:
The race is between `GetRuntimeContextStorage` and `SetRuntimeContextStorage`. Without the lock, we might partially overwrite the `shared_ptr`, thus scrambling up the counter and causing memory corruption.

I went the safer way here. But I'm fine with adding a requirement that context storage customization MUST be done before any other telemetry API call. Then we can remove the lock and win a tiny bit of performance.

Reply:
Agree. there is race condition and possible corruption if `GetRuntimeContextStorage` and `SetRuntimeContextStorage` is called simultaneously as line [163]( https://github.com/pyohannes/opentelemetry-cpp/blob/00cecbb7330ecdcfcefa0050541be41ee32ef78d/api/include/opentelemetry/context/runtime_context.h#L163) is not atomic operation. Although this is tiny bit of performance overhead, it can become significant if `GetRuntimeContextStorage` get's called for each span processing.  Adding a requirement for context storage initialization before telemetry api calls would be good in that case.

Reply:
I removed the locking around `GetRuntimeContextStorage` and `SetRuntimeContextStorage`. I added a note to `SetRuntimeContextStorage` that the behavior is undefined when it it's called after any spans have been created.